crates/ra_prof/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+extern crate lazy_static;
+
+use std::cell::RefCell;
+use std::time;
+use std::fmt;
+use std::mem;
+use std::io::{stderr, StderrLock, Write};
+use std::iter::repeat;
+use std::collections::{HashSet};
+use std::default::Default;
+use std::iter::FromIterator;
+use std::sync::RwLock;
+
+use lazy_static::lazy_static;
+
+type Message = (usize, u64, String);
+
+pub struct Profiler {
+    desc: String,
+}
+
+pub struct Filter {
+    depth: usize,
+    allowed: Vec<String>,
+}
+
+struct ProfileStack {
+    starts: Vec<time::Instant>,
+    messages: Vec<Message>,
+    filter_data: FilterData,
+}
+
+impl ProfileStack {
+    fn new() -> ProfileStack {
+        ProfileStack { starts: Vec::new(), messages: Vec::new(), filter_data: Default::default() }
+    }
+}
+
+#[derive(Default)]
+struct FilterData {
+    depth: usize,
+    version: usize,
+    allowed: HashSet<String>,
+}
+
+impl Clone for FilterData {
+    fn clone(&self) -> FilterData {
+        let set = HashSet::from_iter(self.allowed.iter().cloned());
+        FilterData { depth: self.depth, allowed: set, version: self.version }
+    }
+}
+
+lazy_static! {
+    static ref FILTER: RwLock<FilterData> = RwLock::new(Default::default());
+}
+
+thread_local!(static PROFILE_STACK: RefCell<ProfileStack> = RefCell::new(ProfileStack::new()));
+
+pub fn set_filter(f: Filter) {
+    let mut old = FILTER.write().unwrap();
+    let set = HashSet::from_iter(f.allowed.iter().cloned());
+    let filter_data = FilterData { depth: f.depth, allowed: set, version: old.version + 1 };
+    *old = filter_data;
+}
+
+pub fn profile<T: fmt::Display>(desc: T) -> Profiler {
+    PROFILE_STACK.with(|stack| {
+        let mut stack = stack.borrow_mut();
+        if stack.starts.len() == 0 {
+            match FILTER.try_read() {
+                Ok(f) => {
+                    if f.version > stack.filter_data.version {
+                        stack.filter_data = f.clone();
+                    }
+                }
+                Err(_) => (),
+            };
+        }
+        let desc_str = desc.to_string();
+        if desc_str.is_empty() {
```

Comment:
Why do we need this `is_empty` condition?

Reply:
This is a kind of foolproof. When someone invokes profile method with empty description.